pipeline_attend_and_excite.py

```diff
@@ -1,262 +1,584 @@
+                    " the batch size of `prompt`."
+                )
+            else:
+                uncond_tokens = negative_prompt
+
+            max_length = prompt_embeds.shape[1]
+            uncond_input = self.tokenizer(
+                uncond_tokens,
+                padding="max_length",
+                max_length=max_length,
+                truncation=True,
+                return_tensors="pt",
+            )
+
+            if hasattr(self.text_encoder.config, "use_attention_mask") and self.text_encoder.config.use_attention_mask:
+                attention_mask = uncond_input.attention_mask.to(device)
+            else:
+                attention_mask = None
+
+            negative_prompt_embeds = self.text_encoder(
+                uncond_input.input_ids.to(device),
+                attention_mask=attention_mask,
+            )
+            negative_prompt_embeds = negative_prompt_embeds[0]
+
+        if do_classifier_free_guidance:
+            # duplicate unconditional embeddings for each generation per prompt, using mps friendly method
+            seq_len = negative_prompt_embeds.shape[1]
+
+            negative_prompt_embeds = negative_prompt_embeds.to(dtype=self.text_encoder.dtype, device=device)
+
+            negative_prompt_embeds = negative_prompt_embeds.repeat(1, num_images_per_prompt, 1)
+            negative_prompt_embeds = negative_prompt_embeds.view(batch_size * num_images_per_prompt, seq_len, -1)
+
+            # For classifier free guidance, we need to do two forward passes.
+            # Here we concatenate the unconditional and text embeddings into a single batch
+            # to avoid doing two forward passes
+            prompt_embeds = torch.cat([negative_prompt_embeds, prompt_embeds])
+
+        return text_inputs, prompt_embeds
+
+    @staticmethod
+    def _compute_max_attention_per_index(attention_maps: torch.Tensor,
+                                         indices_to_alter: List[int],
+                                         smooth_attentions: bool = False,
+                                         sigma: float = 0.5,
+                                         kernel_size: int = 3) -> List[torch.Tensor]:
+        """ Computes the maximum attention value for each of the tokens we wish to alter. """
+        attention_for_text = attention_maps[:, :, 1:-1]
+        attention_for_text *= 100
+        attention_for_text = torch.nn.functional.softmax(attention_for_text, dim=-1)
+
+        # Shift indices since we removed the first token
+        indices_to_alter = [index - 1 for index in indices_to_alter]
+
+        # Extract the maximum values
+        max_indices_list = []
+        for i in indices_to_alter:
+            image = attention_for_text[:, :, i]
+            if smooth_attentions:
+                smoothing = GaussianSmoothing(channels=1, kernel_size=kernel_size, sigma=sigma, dim=2).half().cuda()
```

Comment:
I ran the code and got an exception here:
```
RuntimeError: Input type (torch.cuda.FloatTensor) and weight type (torch.cuda.HalfTensor) should be the same
```
This was done when running `run.py` as follows:
```
python run.py --prompt "a cat and a dog" --seeds [0] --token_indices [2,5]
```
Are there any other adjustments you made in order to get this to work? I see that you added `.half()` to the `smoothing` module, but the input should still be a `FloatTensor`, if I'm not mistaken.

Reply:
@AttendAndExcite This can be fixed by casting the tensors to the correct type.

https://github.com/AttendAndExcite/Attend-and-Excite/blob/72fa567a1e3bb3cc1b63cb53a1d9db5fc10b241f/utils/gaussian_smoothing.py#L119

Should be:

```py
return self.conv(input, weight=self.weight.to(input.dtype), groups=self.groups)
```

Reply:
Indeed, after this change it should work smoothly!